engine/src/main/java/org/hibernate/validator/internal/engine/path/NodeImpl.java

```diff
@@ -525,4 +564,113 @@ else if ( !Arrays.equals( parameterTypes, other.parameterTypes ) ) {
 		}
 		return true;
 	}
+
+	boolean isRootPath() {
+		return parent == null && name == null;
+	}
+
+	@Override
+	public Iterator<Path.Node> iterator() {
+		if ( nodes == null ) {
+			nodes = new NodeImpl[size - 1];
+			NodeImpl curr = this;
+			while ( curr.parent != null ) {
+				nodes[curr.size - 2] = curr;
+				curr = curr.parent;
+			}
+		}
+
+		return new NodeIterator( nodes );
```

Comment:
AFAICS, this is not thread safe, which is a problem IMHO.

I think it would be safer to publish a specific full initialized object when we push a `Path` out of the engine. My understanding is that it's only when we build a `ConstraintViolation`?

Or is it more than that?

Reply:
yes, we only need this for the `ConstraintViolation`.... 

yeah I didn't want to synchronize that block .. 
ok let me try something else instead ...